src/libraries/System.Private.CoreLib/src/System/IO/BinaryWriter.cs

```diff
@@ -451,16 +412,69 @@ public virtual void Write(ReadOnlySpan<byte> buffer)
 
         public virtual void Write(ReadOnlySpan<char> chars)
         {
-            byte[] bytes = ArrayPool<byte>.Shared.Rent(_encoding.GetMaxByteCount(chars.Length));
-            try
+            WriteCharsCommonWithoutLengthPrefix(chars, s_writeToBinaryWriter);
+        }
+
+        // dispatches to the virtual BinaryWriter.Write(byte[], int, int) method
+        private static void WriteToBinaryWriter(BinaryWriter sink, byte[] buffer, int offset, int count)
+        {
+            sink.Write(buffer, offset, count);
+        }
+
+        // dispatches to the virtual Stream.Write(byte[], int, int) method
+        private static void WriteToOutStream(BinaryWriter sink, byte[] buffer, int offset, int count)
+        {
+            sink.OutStream.Write(buffer, offset, count);
+        }
+
+        private void WriteCharsCommonWithoutLengthPrefix(ReadOnlySpan<char> chars, WriteBufferToOutputDelegate writeDelegate)
+        {
+            // We'd like to rent a buffer from the array pool if possible, but we don't want to
+            // rent buffers above some upper threshold since they'll be discarded by the array
+            // pool upon return. Additionally, if our input is truly enormous, the call to
+            // GetMaxByteCount might overflow, which we also want to avoid. Technically any
+            // Encoding could expand from chars -> bytes at an enormous ratio and cause us
+            // problems anyway, but this is so unrealistic that we needn't worry about it.
+
+            byte[] rented;
+
+            if (chars.Length <= MaxArrayPoolRentalSize)
             {
-                int bytesWritten = _encoding.GetBytes(chars, bytes);
-                Write(bytes, 0, bytesWritten);
+                // GetByteCount may walk the buffer contents, resulting in 2 passes over the data.
+                // We prefer GetMaxByteCount because it's a constant-time operation.
+
+                int maxByteCount = _encoding.GetMaxByteCount(chars.Length);
+                if (maxByteCount <= MaxArrayPoolRentalSize)
```

Comment:
would it be possible (and worth it) to add a `stackallock` code path for small char arrays? Similar to what you have done for small strings in `Write(string value)`?